### PR TITLE
8264126: Remove TRAPS/THREAD parameter for class loading functions

### PR DESCRIPTION
**Pull request by coleenp**
find_constrained_instance_or_array_klass only passes THREAD so that it can be used in a MutexLocker for SystemDictionary_lock. This can use the MutexLocker that gets Thread::current() without any harm to performance.

The other functions add_loader_constraint, record_linking_constraints, and check_signature_loaders fall out from that.

check_signature_loaders should throw an exception but it unfortunately makes the caller construct the exception message so it doesn't.

Also: is_shared_class_visible{_impl}

Tested with tier1 on 4 Oracle platforms (in progress)

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8264126](https://bugs.openjdk.java.net/browse/JDK-8264126): Remove TRAPS/THREAD parameter for class loading functions

### Reviewers
 * [Calvin Cheung](https://openjdk.java.net/census#ccheung) (@calvinccheung - **Reviewer**) ⚠️ Review applies to 82b984e380380038e746d3d43961af7c2bb45381
 * [Ioi Lam](https://openjdk.java.net/census#iklam) (@iklam - **Reviewer**)
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**)

### Download
To checkout this PR locally:
`$ git fetch https://git.openjdk.java.net/jdk pull/3176/head:pull/3176`
`$ git checkout pull/3176`

To update a local copy of the PR:
`$ git checkout pull/3176`
`$ git pull https://git.openjdk.java.net/jdk pull/3176/head`

### CODE REVIEW

**Review by calvinccheung — APPROVED** [review]
LGTM.

**Comment by coleenp** [comment]
Thanks Calvin!

File: src/hotspot/share/classfile/systemDictionary.cpp
Comment (coleenp) [reply src/hotspot/share/classfile/systemDictionary.cpp:1842]: This will not impact performance though, since we're already taking a lock.

File: src/hotspot/share/ci/ciEnv.cpp
Comment (coleenp) [reply src/hotspot/share/ci/ciEnv.cpp:428]: Ok, that works for ciEnv.cpp.  I'd rather not change the JVMCI code any more and leave it up to the maintainers of that code.  It's just a cleanliness issue not correctness.

File: src/hotspot/share/classfile/systemDictionary.cpp
Comment (coleenp) [reply src/hotspot/share/classfile/systemDictionary.cpp:1234]: ok.

File: src/hotspot/share/classfile/systemDictionaryShared.cpp
Comment (coleenp) [reply src/hotspot/share/classfile/systemDictionaryShared.cpp:1864]: So we always dump dynamic shared spaces from the VMThread?

Comment (iklam) [reply src/hotspot/share/classfile/systemDictionaryShared.cpp]: When DynamicDumpSharedSpaces is true, this code can be executed in a Java thread or a VM thread. When it's in the VM thread, we cannot proceed to the code below and must return immediately.

When DynamicDumpSharedSpaces is false, this code cannot be executed in a VM thread.

Comment (coleenp) [reply src/hotspot/share/classfile/systemDictionaryShared.cpp]: Ok, I think I got the logic right then.

File: src/hotspot/share/ci/ciEnv.cpp
Comment (dholmes-ora) [inline src/hotspot/share/ci/ciEnv.cpp:445]: We could pass current here too.

File: src/hotspot/share/jvmci/jvmciRuntime.cpp
Comment (dholmes-ora) [inline src/hotspot/share/jvmci/jvmciRuntime.cpp:1268]: Can pass THREAD here

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Hi Coleen,

Generally this looks good, but I'm with Ioi here that where we pass through the current thread to avoid needing to manifest Thread::current down the stack (for MutexLockers or ResourceMarks etc) then I would prefer to see that kept - and if necessary relocate the Thread parameter to the beginning. In the past we have consciously added these thread parameters to avoid the Thread::current calls and I prefer not to see that undone.

Thanks,
David

File: src/hotspot/share/ci/ciEnv.cpp
Comment (coleenp) [reply src/hotspot/share/ci/ciEnv.cpp:445]: ok

File: src/hotspot/share/jvmci/jvmciRuntime.cpp
Comment (coleenp) [reply src/hotspot/share/jvmci/jvmciRuntime.cpp:1268]: ok

File: src/hotspot/share/classfile/systemDictionary.cpp
Comment (coleenp) [reply src/hotspot/share/classfile/systemDictionary.cpp:1841]: Ok, I added it back as requested by Ioi and David. It still looks worse and will have no benefit for performance.

**Comment by coleenp** [comment]
Added back a thread parameter and rerunning tier1 sanity tests.

**Review by iklam — APPROVED** [review]

File: src/hotspot/share/classfile/systemDictionaryShared.cpp
Comment (dholmes-ora) [reply src/hotspot/share/classfile/systemDictionaryShared.cpp:1873]: I'm finding the logic change here very hard to follow. Being in the VMThread may imply that DynamicDumpSharedSpaces must be true, but that doesn't necessarily mean that if not in the VMThread then DynamicDumpSharedSpaces must be false.

**Review by dholmes-ora — CHANGES_REQUESTED** [review]
Updates look good - thanks

But I'm having an issue with the systemDictionaryShared change to record_linking_constraints as well.

Thanks,
David

File: src/hotspot/share/classfile/systemDictionaryShared.cpp
Comment (coleenp) [reply src/hotspot/share/classfile/systemDictionaryShared.cpp]: I got this logic wrong, now fixed.

**Review by iklam — APPROVED** [review]

File: src/hotspot/share/classfile/systemDictionaryShared.cpp
Comment (dholmes-ora) [inline src/hotspot/share/classfile/systemDictionaryShared.cpp:1864]: This is still a functional change. If there is a bug and we get here in the VMThread when DynamicDumpSharedSpaces is not set then we will no longer immediately return.

Comment (iklam) [reply src/hotspot/share/classfile/systemDictionaryShared.cpp:1864]: That's fine. In debug build, we will assert both before/after this change. In product build, after this change, some duplicated info will be added via info->record_linking_constraint but it won't hurt.

Comment (dholmes-ora) [reply src/hotspot/share/classfile/systemDictionaryShared.cpp:1864]: Okay. Thanks for clarifying.

**Review by dholmes-ora — APPROVED** [review]
LGTM.

File: src/hotspot/share/classfile/systemDictionaryShared.cpp
Comment (coleenp) [reply src/hotspot/share/classfile/systemDictionaryShared.cpp:1864]: Thanks for looking carefully at this.

**Comment by coleenp** [comment]
Thanks for reviewing.
/integrate

**Pull request closed without merging**
